### PR TITLE
[6.x] FindBy added

### PR DESCRIPTION
**Pull request by ghost**
Sometimes the `find()` function it's not enough, for example, we don't have the id, but have the email of the user to find, it's can be easily use `findBy('email', '=', 'taylor@laravel.com')` or `findBy('email', 'like', '%@laravel.com%')`, this can be done with `where('email', '=', 'taylor@laravel.com')->first()`, but, like the `find()` function, using that function will be to more productivity and will be more easily to find one user with some characteristic.

### CODE REVIEW

**Comment by JosephSilber** [comment]
Noe that `where` works without an operator. If we are indeed to add a `findBy` method, it should also work without providing an operator.

**Comment by afonsogloeden** [comment]
Fixed

**Comment by 36864** [comment]
Collections have a method named `firstWhere` with the same syntax. Should this method have the same name for consistency?

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -351,14 +351,32 @@ public function findMany($ids, $columns = ['*'])
...
+    public function findBy($column, $operatorOrValue = null, $valueOrColumns = null, array $columns = ['*'])
     {
-        return $this->where($column, $operator, $value)->first($columns);
+        $args = func_get_args();
+
+        if (count($args) < 2) {
+            return;
+        }
+
+        if (count($args) == 2) {
```
Comment (thanhtaivtt) [inline src/Illuminate/Database/Eloquent/Builder.php]: I think, you should change `==` to `===` and same with line 371

Comment (netpok) [reply src/Illuminate/Database/Eloquent/Builder.php]: It can be changed, but I don't think it matters in this case

**Comment by netpok** [comment]
If this is merged it would be nice to have have it like the dynamic wheres.

Example:
```php
User::findByName('Joe');
```

**Comment by samdharris** [comment]
Since you've added `findBy`, is it worth also adding `findByOrFail`?

**Comment by deleugpn** [comment]
I'm guessing you're gonna have to write some tests if this is were to be merged.

**Comment by taylorotwell** [comment]
I think you can just use `where` and `first`.

**Pull request closed without merging**